File: 2025_06/leetcode2025-06-28.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<int> maxSubsequence(vector<int>& nums, int k) {
        vector<pair<int, int>> keyVals;
        int n = nums.size();
        for (int i=0; i<n; i++){
            keyVals.emplace_back(nums[i], i);
        }
        sort(keyVals.rbegin(), keyVals.rend());
        vector<pair<int, int>> frontK;
        for (int i=0; i<k; i++){
            frontK.emplace_back(keyVals[i].second, keyVals[i].first);
        }
        sort(frontK.begin(), frontK.end());
        vector<int> res;
        for (auto&kv : frontK){
            res.push_back(kv.second);
        }
        return res;
    }
};

int main(){
    Solution s;
    vector<int> nums = {2,1,3,3};
    int k = 2;
    vector<int> res = s.maxSubsequence(nums, k);
    for (int i : res){
        cout << i << " ";
    }
    cout << endl;
}

File: 2025_06/leetcode2025-06-28.py
from typing import List

class Solution:
    def maxSubsequence(self, nums: List[int], k: int) -> List[int]:
        # 先将nums中的值从大到小排序，我们需要选的就是前k个，但是由于题目要求的是子序列，所以要保证元素的相对顺序不改变，就需要保存每一个元素的下标，再次进行排序，最后取出每一个下标对中的元素即可
        keyVals = [(nums[i], i) for i in range(len(nums))]
        keyVals.sort(reverse = True)
        return [kv[0] for kv in sorted(keyVals[:k], key = lambda x:x[1])]

if __name__ == "__main__":
    s = Solution()
    nums = [2,1,3,3]
    k = 2
    print(s.maxSubsequence(nums, k))
